File: platforms/mesh_host/android/src/main/kotlin/com/frazko/mesh_host/BleLinkHealth.kt
package com.frazko.mesh_host

/** A GATT connection or successful write is not proof of a live remote app. */
internal class BleLinkHealth(private val began: Long = now()) {
    private var lastVerified: Long? = null
    private var lastProbe = began
    fun received(now: Long, authenticated: Boolean) {
        if (authenticated) lastVerified = now
    }
    fun expired(now: Long): Boolean = now - (lastVerified ?: began) >= 12_000
    fun probeDue(now: Long): Boolean {
        if (lastVerified == null || expired(now) || now - lastProbe < 3_000) return false
        lastProbe = now
        return true
    }
    companion object { fun now(): Long = System.nanoTime() / 1_000_000 }
}

File: platforms/mesh_host/ios/mesh_host/Sources/mesh_host/BleLinkHealth.swift
import Foundation

/// Only authenticated inbound traffic proves the peer is still alive.
/// ATT write completion and a cached CoreBluetooth connection do not.
final class BleLinkHealth {
  private let began: TimeInterval
  private var lastVerified: TimeInterval?
  private var lastProbe: TimeInterval
  init(now: TimeInterval = ProcessInfo.processInfo.systemUptime) {
    began = now
    lastProbe = now
  }
  func received(now: TimeInterval, authenticated: Bool) {
    if authenticated { lastVerified = now }
  }
  func expired(now: TimeInterval) -> Bool { now - (lastVerified ?? began) >= 12 }
  func probeDue(now: TimeInterval) -> Bool {
    guard lastVerified != nil, !expired(now: now), now - lastProbe >= 3 else { return false }
    lastProbe = now
    return true
  }
}
